Haskell Parser Combinators/article.md

```diff
@@ -0,0 +1,1330 @@
+
+<details>
+  <summary>Solution</summary>
+One possible solution using `<|>`:
+
+```hs
+integer :: Parser Integer
+integer = label "integer" $ read <$> (some numberChar <|> ((:) <$> char '-' <*> some numberChar))
+```
+
+However, later in the tutorial, we will see an even better way, using Megaparsec's own mechanisms.
+  <hr>
+</details>
+
+### Strings
+
+To parse a string, the idea is to find everything that is between double-quotes. This approach has a weakness, though: we won't be able to parse escaped quotes. Later in this tutorial, we will present a solution to escaped characters.
+
+To achieve this, we will use two new functions: `between` and `takeWhileP`. Using `between open close x` is equivalent to `open *> x <* close`, while `takeWhileP` will read as many characters as possible that satisfy some predicate.
+
+Note that `takeWhileP` also accepts, optionally, a label for each character that it tries to read. For now, we will leave it as `Nothing`.
+
+```hs
+str :: Parser String
+str = label "string" $ between (char '"') (char '"') (takeWhileP Nothing (/= '"'))
+-- Or:
+--- str = label "string" $ char '"' *> takeWhileP Nothing (/= '"') <* char '"'
+```
+
+In `atom`, include `SString <$> str`.
+
+```hs
+>>> parseTest atom "\"hey\""
+SString "hey"
+```
+
+### Identifiers
+
+An identifier will be like a variable in a programming language. We can choose the naming convention for such a thing, but here I chose the ones that are frequently used for C-like languages: the first letter must be a letter or an underscore, while the remainder may be a letter, digit, or underscore.
+
+Here we use the `many` function, which is very similar to the `some` function discussed before. The difference is that `any` tries to run the given parser **0 or more times**. In Regex, our `identifier` parser is equivalent to `[a-zA-Z][a-zA-Z0-9]*`.
+
+```hs
+identifier :: Parser Identifier
+identifier = label "identifier" $ do
+  first <- letterChar <|> char '_'
+  rest <- many $ alphaNumChar <|> char '_'
+  pure $ Identifier $ first : rest
+-- Or:
+-- identifier = Identifier <$> label "identifier" ((:) <$> (letterChar <|> char '_') <*> many (alphaNumChar <|> char '_'))
+```
+
+Now include `SId <$> identifier` to `atom`.
+
+**Warning**: When parsing with `<|>` or `choice`, **the order of the parsers matters**. Take this as an example:
+
+```hs
+>>> parseTest (SBool <$> bool <|> SId <$> identifier) "true"
+SBool True
+
+>>> parseTest (SId <$> identifier <|> SBool <$> bool) "true"
+SId (Identifier {getId = "true"})
+```
+
+In the first case, we parse a boolean before an identifier, which causes `true` to be matched by `bool`. However, in the second example, using `identifier` before `bool` caused `true` to be recognized as an identifier. You may assume that parsers are greedy, and you should be careful while considering such things. Notice that in the second case, the `bool` parser will never be executed.
```

Comment:
Second sentence is a bit awkward 

You may assume that parsers are greedy, and you should be careful while considering such things.

Such things kind of point to the first clause of the sentence (since they are connected), and it raises confusion since I assume you mean such things = order of parsers. 

Perhaps it can go like this:

You should be careful while considering such things. In most cases, you can assume that parsers are greedy. Notice that in the second case..

Reply:
Reworded it.